[Merged by Bors] - feat: more extensions for differential geometry elaborators

grunweg opened the pull request:
Support inferring a model with corners in all remaining instances in mathlib, except for products and direct sums:
- on Euclidean space, half-space or quadrants.
- for the units in a normed algebra (including GL(V))
- the unit interval
- on the complex unit circle
- on a metric sphere in a real or complex inner product space
- also support finding a model on a domain which is an `InnerProductSpace` (hence only indirectly a normed space)

---

- [x] depends on: #30413
- [x] depends on: #30879
- [x] depends on: #33923 (a 5 line fix)

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include at least one commit authored by each
co-author among the commits in the pull request. If necessary, you may 
create empty commits to indicate co-authorship, using commands like so:

git commit --author="Author Name <author@email.com>" --allow-empty -m "add Author Name as coauthor"

When merging, all the commits will be squashed into a single commit listing all co-authors.

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> List.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
Support Euclidean half-space (with tests)

grunweg committed:
Also support Euclidean space and Euclidean quadrants

grunweg committed:
Infer the models on units of a normed algebra (and fix a typo-ed section name)

grunweg committed:
chore: extract test for clms
This function needs a better name!

grunweg committed:
wip/stub: infer the sphere in a normed space

grunweg committed:
Tweak and comment, further progress

grunweg committed:
fix: Metric.sphere is a coerced set, not a type!

grunweg committed:
wip fix: need to search for an inner product space isntead!

grunweg committed:
Stuck with a qq/syntax quoting problem now

grunweg committed:
chore: make doc-string more complete

grunweg committed:
feat: infer the model on unitInterval

grunweg committed:
Progress on the inferring the model on a sphere: my test still fails...

grunweg committed:
wip: add support for Circle

grunweg committed:
Add tests and make them pass.

grunweg committed:
Fix the generation for spheres, at least

grunweg committed:
feat: implement the last part of the sphere inference

Update tests (and add a new one): so progress, but not good enough.
The mathematical tests also need fixing first.

grunweg committed:
playing around, not solution yet

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -7,7 +7,7 @@ module
 public import Mathlib.Geometry.Manifold.ContMDiff.Defs
 public import Mathlib.Geometry.Manifold.MFDeriv.Defs
-
+public import Mathlib.Analysis.InnerProductSpace.Defs -- temporary
```
👀?

thorimur commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -406,6 +440,143 @@ where
+                  return some n
+                | _ =>
+                  throwError "found a fact about `finrank ℝ E`, but the right hand \
+                    side `{rhs}` is not of the form `m + 1` for some `m`"
```
I'm not familiar with the library here, so maybe this is silly, but: could we just return something like `rhs - 1`, and see what happens? Do we ever encounter `0 - 1 = 1`? Or, are these rhs's basically always of the form `n + 1`, so that this matching is more reliable than it might seem on first glance? 😁

thorimur replied:
Also: is there a chance that the rhs is not a literal natural number at all? Do we want it to be a literal natural number?

If so—I wonder if it would make sense to call `norm_num` internals on the result, which iirc can detect that sort of thing 👀

grunweg committed:
Merge branch 'master' into diffgeo-custom-elaborators-evenmore

grunweg committed:
Move sphere tests to a new file, and completely revamp them

grunweg committed:
Small tweaks, and another test

grunweg committed:
Avoid importing InnerProductSpace

grunweg committed:
Better errors for CLM, and tweak transparency check

grunweg committed:
Restore newline

grunweg committed:
Comment on searching for NormedAlgebra instances

grunweg committed:
feat: also search for the self-model on an inner product space

grunweg committed:
chore: add test for inner product space

grunweg committed:
fixup tests, part 1

grunweg committed:
wip: found the defect; will incoming after the meeting!

grunweg committed:
Merge branch 'master' into diffgeo-custom-elaborators-evenmore

grunweg committed:
fix the bug; new tests pass indeed

grunweg committed:
update Test output

joneugster reviewed (commented):
I'm not that familiar with the mathematics behind this meta-PR, but I think it makes sense and the additions to `findModel` seem reasonable and are well-structured!

I've just had a question about the imports of this file, but there might be no action needed, as you see fit.

grunweg committed:
Clarify comment

grunweg committed:
Apply suggestions from code review

Co-authored-by: Thomas R. Murrills <68410468+thorimur@users.noreply.github.com>

Vierkantor commented on Mathlib/Geometry/Manifold/Notation.lean:
```diff
@@ -336,6 +359,20 @@ where
+    let some (inst, K) ← findSomeLocalInstanceOf? `InnerProductSpace fun inst type ↦ do
+      -- We don't use `match_expr` here to avoid importing `InnerProductSpace`.
+      match (← instantiateMVars type).cleanupAnnotations with
+        | mkApp4 (.const `InnerProductSpace _) k E _ _ =>
```
This seems quite fragile: what if `InnerProductSpace` gets a different number of arguments? Wouldn't [Lean.Expr.withApp](https://leanprover-community.github.io/mathlib4_docs/Lean/Expr.html#Lean.Expr.withApp) + ``Expr.constName? == some `InnerProductSpace`` be more robust?

Vierkantor reviewed (approved):
bors d+

grunweg committed:
chore: add a test for ContinuousLinearMaps over not the identity

This code has not been changed in this PR, but let's add a test anyway.
Any behaviour or diagnostics changes will come in a future PR.

grunweg committed:
Apply suggestion from @Vierkantor

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

pre-commit-ci-lite[bot] committed:
[pre-commit.ci lite] apply automatic fixes

grunweg committed:
Fix line length; update test output

grunweg committed:
Disable brittle test
